src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs

```diff
@@ -0,0 +1,517 @@
+                }
+
+                if (parameter.Constraints.Count > 0)
+                {
+                    parameterConstraints.AddRange(parameter.Constraints);
+                }
+
+                return ParameterPartCore(
+                    parameter.Name,
+                    @default,
+                    parameter.ParameterKind,
+                    (IEnumerable<RoutePatternConstraintReference>)parameterConstraints ?? Array.Empty<RoutePatternConstraintReference>());
+            }
+        }
+
+        public static RoutePatternPathSegment Segment(IEnumerable<RoutePatternPart> parts)
+        {
+            if (parts == null)
+            {
+                throw new ArgumentNullException(nameof(parts));
+            }
+
+            return SegmentCore(parts);
+        }
+
+        public static RoutePatternPathSegment Segment(params RoutePatternPart[] parts)
+        {
+            if (parts == null)
+            {
+                throw new ArgumentNullException(nameof(parts));
+            }
+
+            return SegmentCore(parts);
+        }
+
+        private static RoutePatternPathSegment SegmentCore(IEnumerable<RoutePatternPart> parts)
+        {
+            return new RoutePatternPathSegment(parts.ToArray());
+        }
+
+        public static RoutePatternLiteralPart LiteralPart(string content)
+        {
+            if (string.IsNullOrEmpty(content))
+            {
+                throw new ArgumentException(Resources.Argument_NullOrEmpty, nameof(content));
+            }
+
+            if (content.IndexOf('?') >= 0)
+            {
+                throw new ArgumentException(Resources.FormatTemplateRoute_InvalidLiteral(content));
+            }
+
+            return LiteralPartCore(content);
+        }
+
+        private static RoutePatternLiteralPart LiteralPartCore(string content)
+        {
+            return new RoutePatternLiteralPart(content);
+        }
+
+        public static RoutePatternSeparatorPart SeparatorPart(string content)
+        {
+            if (string.IsNullOrEmpty(content))
+            {
+                throw new ArgumentException(Resources.Argument_NullOrEmpty, nameof(content));
+            }
+
+            return SeparatorPartCore(content);
+        }
+
+        private static RoutePatternSeparatorPart SeparatorPartCore( string content)
+        {
+            return new RoutePatternSeparatorPart( content);
+        }
+
+        public static RoutePatternParameterPart ParameterPart(string parameterName)
+        {
+            if (string.IsNullOrEmpty(parameterName))
+            {
+                throw new ArgumentException(Resources.Argument_NullOrEmpty, nameof(parameterName));
+            }
+
+            if (parameterName.IndexOfAny(RoutePatternParser.InvalidParameterNameChars) >= 0)
+            {
+                throw new ArgumentException(Resources.FormatTemplateRoute_InvalidParameterName(parameterName));
```

Comment:
`if (string.IsNullOrEmpty(parameterName))` and `if (parameterName.IndexOfAny(RoutePatternParser.InvalidParameterNameChars) >= 0)` are checked in `ParameterPartCore` and all its `ParameterPart` callers.

Don't need to duplicate check.

Reply:
We generally put all argument validation as close as possible to the method the user called (see null checks all over the codebase).

Reply:
In that case you could leave this check (and in other overloads) and remove the one in ParameterPartCore.